Federation tutorial (#29)
spiffe/spire-tutorials: federation/README.md

> 

Author: I see...in favor of keeping the domains used in the tutorial I would say that SPIFFE trust domains are not tied nor related to DNS domain names...despite being convenient to use the same trust domain as the owned domain name.
Let's say we want to change this, do you believe stockmarket.example and broker.example are ok?

> 

Reviewer: I had kind of forgotten that trust domains are not tied to DNS names. Based on that, it does weigh the decision more in favor of leaving them as is. But yes, stockmarket.example and broker.example are ok to use. On the other hand, I do think we want to change stockSmarket.org, so maybe go for the *.example forms.

> 

Author: I will take the chance to fix both, is more or less the same effort. Thank you!

# Overview

This tutorial shows how to authenticate two SPIFFE-identified workloads that are identified by two different SPIRE Servers.

In this tutorial you will learn how to:

> * Configure each SPIRE Server to expose it's SPIFFE Federation bundle endpoint using SPIFFE authentication and WebPKI authentication.

Author: Obviously you are absolutely right. I usually use "its" properly...but not this time 🤦‍♂️ Thanks!

# Prerequisites

* Two SPIRE Servers instances.
* Two SPIRE Agents, one connected to one SPIRE Server, and the second connected to the other SPIRE Server.
* Two workloads that needs to communicate each other via mTLS, and use the Workload API to get SVIDs and trust bundles.

# Scenario

Let's say we have a stock broker's webapp that wants to display stock quotes fetched from a stocks market webservice provider. The scenario goes as follows:  

1. The user enters the broker's webapp stock quotes URL in a browser.
2. The webapp workload receives the request and makes an HTTP request for quotes to the stocks market service using mTLS.
3. The stocks market service receives the request and sends the quotes in the response.
4. The webapp renders the stock quotes page using the returned quotes and sends it to the browser.
5. The browser displays the quotes to the user. The webapp includes some javascript to refresh the page every 1 second, so every second these steps are executed again.

In addition to the above and for the rest of this tutorial, we are going to assume the following trust domain names for their SPIRE installations: `broker.org` and `stocksmarket.org`.  

> Also, the applications access the WorkloadAPI directly to get SVIDs and trust bundles, meaning there are no proxies in the scenario described.

Author: Sure, I'm ok, I'll just find & replace the string. Thank you.

> # Configuring SPIFFE Federation endpoint

Author: The issue is that it is much easier to see federation in action using SPIFFE auth in both sides, because we can do it by just using docker-compose, without requiring to have ownership over any domain name.

> # Configuring SPIFFE Federation endpoint

Reviewer: Ah, so the docker-compose demo only does the SPIFFE auth, but you are also explaining the alternate method using Web PKI. This makes sense, so let's keep the existing structure. I think we should just explicitly explain that in the doc. So in the first Web PKI section, insert a new first sentence in the first paragraph. Something like:

Although not shown in the Docker Compose tutorial, you could make the changes shown here to use Web PKI for authentication instead of SPIFFE auth.

> # Configuring SPIFFE Federation endpoint

Author: What if I add something like that under the `Federation Example Using SPIFFE Authentication with SPIRE 0.11.0` H1?

> # Configuring SPIFFE Federation endpoint

Reviewer: Sure, you can put it in the explanation of the docker-compose demo as you suggest.

Then, to configure the broker's SPIRE Server bundle endpoint, we configure the `federation` section as follows:

```hcl
server {
    .
    .
    trust_domain = "broker.org"
    .
    .

    federation {
        bundle_endpoint {
            address = "0.0.0.0"
            port = 443
            acme {
                domain_name = "broker.org"
                email = "some@email.com"
                tos_accepted = true
            }   
        }
    }
}
```
This will publish the federation endpoint in any IP address at port 443. We use port 443 because we will be using Let's Encrypt as our ACME provider (this is used by default, if you want to use a different one then you must set the `directory_url` configurable). Note that `tos_accepted` was set to `true`, meaning that we accept the terms of service of our ACME provider, which in turn is needed when using Let's Encrypt.

Worth noting that for this to work you must own the `broker.org` domain and it must point to the SPIRE Server exposing the federation bundle endpoint.

> # Configure the SPIRE Servers to retrieve trust bundles from each other

Author: > I have the same comment about simplifying the explanation so we have a single scenario with one showing SPIFFE auth and the other showing Web PKI auth.

Same answer. It is easier to use SPIFFE auth in both sides to see federation in action.

> # Configure the SPIRE Servers to retrieve trust bundles from each other

Reviewer: Same answer here, too 😄  - keep it as is.

This is the other part of the SPIRE Server configuration needed in order to achieve federation, we need to tell each SPIRE Server where it can find the trust bundles for other trust domains, in other words, we have to configure the `federate_with` section. The configuration of this section has some slight differences when using the different methods of authentication.

## Using SPIFFE authentication

> As we saw previously, the SPIRE Server of the stocks market service provider has its federation endpoint listening on port `8443` at any IP. We will also assume that `spire-server-stocks` is a DNS name that resolves to the stocks market service's SPIRE Server IP. Then, the broker's SPIRE Server must be configured with the following `federate_with` section:

Author: No, it is not, because `stockmarket.org` is a SPIFFE trust domain, not a domain name. The actual host has the `spire-server-stocks` name as configured in the docker-compose.yml file.

> As we saw previously, the SPIRE Server of the stocks market service provider has its federation endpoint listening on port `8443` at any IP. We will also assume that `spire-server-stocks` is a DNS name that resolves to the stocks market service's SPIRE Server IP. Then, the broker's SPIRE Server must be configured with the following `federate_with` section:

Reviewer: OK, I understand. Please do use the phrase "IP address" though, rather than just IP.

> As we saw previously, the SPIRE Server of the stocks market service provider has its federation endpoint listening on port `8443` at any IP. We will also assume that `spire-server-stocks` is a DNS name that resolves to the stocks market service's SPIRE Server IP. Then, the broker's SPIRE Server must be configured with the following `federate_with` section:

Author: Oh yeah, I fixed that. Thanks

        bundle_endpoint {
            address = "0.0.0.0"
            port = 8443
        }
        federates_with "broker.org" {
            bundle_endpoint {
                address = "broker.org"
                use_web_pki = true
            }
        }
    }
}
```
The differences are:
- `port` was removed. This is because by default it is set to `443`, which is the port where the broker's federation bundle endpoint is listening.
- `address` now is set to the broker's domain `broker.org`.
- `use_web_pki` was added and set to `true`. This is mandatory when the bundle endpoint to which we want to federate with is using WebPKI authentication.

# Bootstrap federation

We have configured the SPIRE Servers with the address of the federation endpoints, but this is not enough to make federation work. To enable the SPIRE Servers to fetch the trust bundles from each other they need each other's trust bundle first, because they have to authenticate the SPIFFE identity of the federated server that is trying to access the federation endpoint. Once federation is bootstrapped the trust bundle updates are fetched troughth the federation endpoint API using the current trust bundle.

> This is done using a couple of SPIRE Server commands: `experimental bundle show` and `experimental bundle set`.

Reviewer 2: The `experimental bundle` commands have been subsumed into the corresponding `bundle` commands using the `-format spiffe` flag in SPIRE v0.11.0. The experimental commands will be removed in a future release so we should avoid using them in this tutorial.
E.g.: `spire-server bundle show -format spiffe > broker.org.bundle` is the equivalent to `spire-server experimental bundle show > broker.org.bundle`

> This is done using a couple of SPIRE Server commands: `experimental bundle show` and `experimental bundle set`.

Author: oh yeah! forgot about this! thank you!

```

At this point the stocks market service's SPIRE Server is able to validate SVIDs having SPIFFE IDs with a `broker.org` trust domain. However, the broker's SPIRE Server is not yet able to validate SVIDs having SPIFFE IDs with a `stocksmarket.org` trust domain. To make this possible, the stocks market folks must run the following on the node they have SPIRE Server running:

```
stocks-market> spire-server experimental bundle show > stocksmarket.org.bundle
```

Then the stocks market folks must give a copy of this file to the broker folks, so they can store this trust bundle on their SPIRE Server and associate it with the `stocksmarket.org` trust domain. To achieve this, the broker folks must run the following on the node where they have SPIRE Server running:

```
broker> spire-server experimental bundle set -id spiffe://stocksmarket.org -path /some/path/stocksmarket.org.bundle
```

Now both SPIRE Servers can validate SVIDs having SPIFFE IDs with each other's trust domain, thus both can start fetching trust bundle updates from each other's federation endpoints. Also, as of now they can create registration entries for federating.

> Something important to consider is that the creation of the `broker.org.bundle` file (and later setting by the stocks market service) is not needed when the broker's SPIRE Server is using WebPKI authentication for its federation bundle endpoint.

Author: I believe it is, and I like it.

# Create registration entries for federation

Now that they have each other trust bundle, let's see how they can create registration entries to federate with each other.

To simplify things, we are going to suppose that the webapp and the quotes service are running both on Linux boxes, one owned by the broker and the other owned by the stocks market organization. Since they are using SPIRE, each Linux box also has a SPIRE Agent installed. In addition to this, the webapp is run using the `webapp` user, and the quotes service is run using the `quotes-service` user.

Then, in the SPIRE Server node of the broker, they must create a registration entry as follows:

```
broker> spire-server entry create \
	-parentID <SPIRE Agent's SPIFFE ID> \
	-spiffeID spiffe://broker.org/webapp \
	-selector unix:user:webapp \
	-federatesWith "spiffe://stocksmarket.org"
```

> Once this registration entry is created, when the webapp asks for an SVID it will get one having the `spiffe://broker.org/webapp` identity, along with the trust bundle associated to the `stocksmarket.org` trust domain, due to the use of the `-federateWith` flag when creating the registration entry. Worth to mention that if the `-federateWith` flag would not be used, federation would not work.

Author: Makes sense to me.

At the stocks market service side, they must create a registration entry as follows:

```
stocks-market> spire-server entry create \
	-parentID <SPIRE Agent's SPIFFE ID> \
	-spiffeID spiffe://stocksmarket.org/quotes-service \
	-selector unix:user:quotes-service \
	-federatesWith "spiffe://broker.org"
```

Similarly, once this registration entry is created, when the quotes service asks for an SVID it will get one having the `spiffe://stocksmarket.org/quotes-service` identity, along with the trust bundle associated to the `broker.org` trust domain.

That is about it, now all the pieces are together to make federation work and see how the webapp is able to communicate with the quotes service despite having identities with different trust domains.

# Federation Example using SPIFFE authentication with SPIRE 0.11.0

This section explains how to try the example implementation of the scenario described previously.

> ## Requirements
> 
> - Go 1.14
> - docker-compose

Reviewer: I copied this from Andrés' nested SPIRE tutorial. Is the new, first bullet correct?
```suggestion
## Requirements

Required files for this tutorial can be found in the `nested-spire` directory in https://github.com/spiffe/spire-tutorials. If you didn't already clone the repository please do so now.

Before proceeding, review the following system requirements:
- A 64-bit Linux or macOS environment
- [Docker](https://docs.docker.com/get-docker/) and [Docker Compose](https://docs.docker.com/compose/install/) installed (Docker Compose is included in macOS Docker Desktop)
- [Go](https://golang.org/dl/) 1.14.4 or higher installed
```

Author: Yes, it is correct, I will add it. Thanks!

## Build

> 

Reviewer: I'm guessing here. Please review it carefully.

```suggestion
Ensure that the current working directory is `.../spire-tutorials/federation` and run the following command to create the files needed for Docker Compose:
```

```
$ ./build.sh
```

## Run

> 
> ```
> $ docker-compose up -d
> ```
> 
> This starts the SPIRE Servers and the applications.

Reviewer: Please move that sentence up a few lines as shown. I had to escape the backticks in this GitHub comment. Please ignore that.

```suggestion

Run the following command to start the SPIRE Servers and the applications:

\`\`\`
$ docker-compose up -d
\`\`\`
```

## Start SPIRE Agents 

> 

Reviewer: ```suggestion
Run the following command to start the SPIRE Agents:
```

```
$ ./1-start-spire-agents.sh
```

## Bootstrap Federation

> 

Reviewer: ```suggestion
Run the following command to [bootstrap the federation](#bootstrap-federation):
```

```
$ ./2-bootstrap-federation.sh
```

## Create Workload Registration Entries

> 

Reviewer: ```suggestion
Run the following command to create [workload registration entries](#create-registration-entries-for-federation):
```

```
$ ./3-create-registration-entries.sh
```

After running this script, it may take some seconds for the applications to receive their SVIDs and trust bundles. 

> ## See it working on the browser

Reviewer: ```suggestion
## See the Scenario Working In a Browser
```

Open up a browser to http://localhost:8080/quotes and you should see a grid of randomly generated phony stock quotes that are updated every 1 second.

> 

Reviewer: If you have time, it would be nice to include commands to cat the federation stanzas in the server.conf files, show the registration entries, and maybe more.

> ## Clean up

Reviewer: "Clean up your mess" would be correct. Here, we're describing a task: cleanup.
```suggestion
## Cleanup
```

    }
}
```
This will publish the federation endpoint in any IP address at port 443. We use port 443 because we will be using Let's Encrypt as our ACME provider (this is used by default, if you want to use a different one then you must set the `directory_url` configurable). Note that `tos_accepted` was set to `true`, meaning that we accept the terms of service of our ACME provider, which in turn is needed when using Let's Encrypt.

For SPIFFE Federation using Web PKI to work, you must own the domain specified for `domain_name` (`broker.example` in our example) and the domain must resolve to the SPIRE Server exposing the federation bundle endpoint.

# Configure SPIRE Servers to Retrieve Trust Bundles From Each Other

After configuring federation endpoints, the next step to enable SPIFFE federation is to configure the SPIRE Servers to find the trust bundles for other trust domains. The `federates_with` configuration option in `server.conf` is where you specify the endpoint of the other trust domain. The configuration of this section has some slight differences when using the different methods of authentication.

## Configure Trust Bundle Location Using SPIFFE Authentication

> As we saw previously, the SPIRE Server of the stock market service provider has its federation endpoint listening on port `8443` at any IP address. We will also assume that `spire-server-stock` is a DNS name that resolves to the stock market service's SPIRE Server IP address. Then, the broker's SPIRE Server must be configured with the following `federate_with` section:

Author: What I meant is that for the docker compose example we don't have to use a FQDN there, instead we just need the hostname because all the containers are running on the same network. But, in a more realistic scenario it would likely be a FQDN.

> As we saw previously, the SPIRE Server of the stock market service provider has its federation endpoint listening on port `8443` at any IP address. We will also assume that `spire-server-stock` is a DNS name that resolves to the stock market service's SPIRE Server IP address. Then, the broker's SPIRE Server must be configured with the following `federate_with` section:

Reviewer: Ah, so you **do** mean "DNS name". It might be good to clarify this for users. See if this sentence makes sense. Feel free to suggest changes.
```suggestion
As we saw previously, the SPIRE Server of the stock market service provider has its federation endpoint listening on port `8443` at any IP address. We will also assume that `spire-server-stock` is a DNS name that resolves to the stock market service's SPIRE Server IP address. (The Docker Compose demo here uses the hostname `spire-server-stock`, but in typical usage you would specify a FQDN.) Then, the broker's SPIRE Server must be configured with the following `federate_with` section:
```

- `address` now is set to the broker's domain `broker.example`.
- `use_web_pki` was added and set to `true`. This is mandatory when the bundle endpoint to which we want to federate is using Web PKI authentication.

# Bootstrap Federation

We have configured the SPIRE Servers with the address of the federation endpoints, but this is not enough to make federation work. To enable the SPIRE Servers to fetch the trust bundles from each other they need each other's trust bundle first, because they have to authenticate the SPIFFE identity of the federated server that is trying to access the federation endpoint. Once federation is bootstrapped, the trust bundle updates are fetched trough the federation endpoint API using the current trust bundle.

The bootstrapping is done using a couple of SPIRE Server commands: `bundle show` and `bundle set`.

## Get the Bootstrap Trust Bundle

Let's say we want to get the broker's SPIRE Server trust bundle. On the node where the SPIRE Server is running we run:

```
broker> spire-server bundle show -format spiffe > broker.example.bundle
```

> This saves the trust bundle in the `broker.og.bundle` file. Then the broker must give a copy of this file to the stock market service folks, so they can store this trust bundle on their SPIRE Server and associate it with the `broker.example` trust domain. To achieve this, the stock market service folks must run the following on the node where they have SPIRE Server running:

Author: Woah, good catch!

```

By specifying the `-federatesWith` flag, once this registration entry is created, when the webapp's SPIRE Server asks for an SVID it will get one from the broker's SPIRE Server with the `spiffe://broker.example/webapp` identity, along with the trust bundle associated to the `stockmarket.example` trust domain.

On the stock market service side, they must create a registration entry as follows:

```
stock-market> spire-server entry create \
	-parentID <SPIRE Agent's SPIFFE ID> \
	-spiffeID spiffe://stockmarket.example/quotes-service \
	-selector unix:user:quotes-service \
	-federatesWith "spiffe://broker.example"
```

Similarly, once this registration entry is created, when the quotes service asks for an SVID it will get one having the `spiffe://stockmarket.example/quotes-service` identity, along with the trust bundle associated to the `broker.example` trust domain.

That is about it. Now all the pieces are in place to make federation work and demonstrate how the webapp is able to communicate with the quotes service despite having identities with different trust domains.

# Federation Example Using SPIFFE Authentication with SPIRE 0.11.0

> This section explains how to use Docker Compose to try an example implementation of one of the scenarios described in this tutorial.
> 
> Although not shown here, you could make the changes shown in the Web PKI authentication sections to try the Web PKI scenario.

Reviewer: Please verify the new last sentence.

```suggestion
This section explains how to use Docker Compose to try an example implementation of the SPIFFE auth scenario described in this tutorial.

Although not shown here, you could make the changes shown in the Web PKI authentication sections to try the Web PKI scenario. Remember that to configure Web PKI, the FQDN specified for `domain_name` must be owned by you and resolvable over the internet via DNS. 
```

- [Go](https://golang.org/dl/) 1.14.4 or higher installed

## Build

Ensure that the current working directory is `.../spire-tutorials/federation` and run the following command to create the files needed for Docker Compose:

```
$ ./build.sh
```

## Run

Run the following command to start the SPIRE Servers and the applications:

```
$ docker-compose up -d
```

## Start SPIRE Agents 

Run the following command to start the SPIRE Agents:

```
$ ./1-start-spire-agents.sh
```

## Bootstrap Federation

Run the following command to [bootstrap the federation](#bootstrap-federation):

```
$ ./2-bootstrap-federation.sh
```

## Create Workload Registration Entries

Run the following command to create [workload registration entries](#create-registration-entries-for-federation):

```
$ ./3-create-registration-entries.sh
```

After running this script, it may take some seconds for the applications to receive their SVIDs and trust bundles. 

## See the Scenario Working In a Browser

Open up a browser to http://localhost:8080/quotes and you should see a grid of randomly generated phony stock quotes that are updated every 1 second.

## See the Configuration

> Too see the broker's SPIRE Server configuration you can run:

Reviewer: ```suggestion
To see the broker's SPIRE Server configuration you can run:
```

You should see:

```
server {
    bind_address = "0.0.0.0"
    bind_port = "8081"
    registration_uds_path = "/tmp/spire-registration.sock"
    trust_domain = "broker.example"
    data_dir = "/opt/spire/data/server"
    log_level = "DEBUG"
    log_file = "/opt/spire/server.log"
    default_svid_ttl = "1h"
    ca_subject = {
        country = ["US"],
        organization = ["SPIFFE"],
        common_name = "",
    }

    federation {
        bundle_endpoint {
            address = "0.0.0.0"
            port = 8443
        }
        federates_with "stockmarket.example" {
            bundle_endpoint {
                address = "spire-server-stock"
                port = 8443
            }
        }    
    }
}

plugins {
    DataStore "sql" {
        plugin_data {
            database_type = "sqlite3"
            connection_string = "/opt/spire/data/server/datastore.sqlite3"
        }
    }

        NodeAttestor "x509pop" {
                plugin_data {
                        ca_bundle_path = "/opt/spire/conf/server/agent-cacert.pem"
                }
        }

    NodeResolver "noop" {
        plugin_data {}
    }

    KeyManager "memory" {
        plugin_data = {}
    }
}
```

> Too see the stock market's SPIRE Server configuration you can run:

Reviewer: ```suggestion
To see the stock market's SPIRE Server configuration you can run:
```
